File: php/src/Browser.php
<?php

declare(strict_types=1);

namespace Rustwright;

final class Browser
{
    private bool $closed = false;
    private bool $freed = false;

    /** @var list<\WeakReference<Page>> */
    private array $pages = [];

    public function __construct(
        private readonly NativeLibrary $native,
        private mixed $handle,
    ) {
    }

    public function newPage(): Page
    {
        $this->ensureOpen();
        $page = new Page($this->native, $this->native->browserNewPage($this->handle));
        $this->pages[] = \WeakReference::create($page);
        return $page;
    }

    public function wsEndpoint(): string
    {
        $this->ensureOpen();
        return $this->native->browserWsEndpoint($this->handle);
    }

    public function close(): void
    {
        if ($this->freed) {
            return;
        }

        $firstError = null;
        foreach ($this->pages as $reference) {
            $page = $reference->get();
            if ($page === null) {
                continue;
            }
            try {
                $page->close();
            } catch (\Throwable $error) {
                $firstError ??= $error;
            }
        }
        $this->pages = [];

        try {
            if (!$this->closed) {
                $this->native->browserClose($this->handle);
                $this->closed = true;
            }
        } catch (\Throwable $error) {
            $firstError ??= $error;
        } finally {
            $this->native->browserFree($this->handle);
            $this->handle = null;
            $this->freed = true;
        }

        if ($firstError !== null) {
            throw $firstError;
        }
    }

    public function __destruct()
    {
        try {
            $this->close();
        } catch (\Throwable) {
            // Destructors cannot safely surface lifecycle errors.
        }
    }

    private function ensureOpen(): void
    {
        if ($this->closed || $this->freed) {
            throw new RustwrightException('Browser is closed');
        }
    }
}

File: java/src/main/java/com/skyvern/rustwright/Browser.java
package com.skyvern.rustwright;

import java.lang.foreign.MemorySegment;
import java.util.ArrayList;
import java.util.Collections;
import java.util.IdentityHashMap;
import java.util.Set;

/** An owned RwBrowser handle. close() closes pages and frees every native handle once. */
public final class Browser implements AutoCloseable {
    private final NativeBindings bindings;
    private final Object lock = new Object();
    private final Set<Page> pages = Collections.newSetFromMap(new IdentityHashMap<>());
    private MemorySegment handle;
    private boolean freed;

    Browser(NativeBindings bindings, MemorySegment handle) {
        this.bindings = bindings;
        this.handle = handle;
    }

    public Page newPage() {
        synchronized (lock) {
            requireOpen();
            Page page = new Page(this, bindings, bindings.browserNewPage(handle), lock);
            pages.add(page);
            return page;
        }
    }

    public String wsEndpoint() {
        synchronized (lock) {
            requireOpen();
            return bindings.browserWsEndpoint(handle);
        }
    }

    @Override
    public void close() {
        synchronized (lock) {
            if (freed) {
                return;
            }

            RuntimeException failure = null;
            for (Page page : new ArrayList<>(pages)) {
                try {
                    page.close();
                } catch (RuntimeException error) {
                    failure = combine(failure, error);
                }
            }
            try {
                bindings.browserClose(handle);
            } catch (RuntimeException error) {
                failure = combine(failure, error);
            }
            try {
                bindings.browserFree(handle);
            } catch (RuntimeException error) {
                failure = combine(failure, error);
            } finally {
                handle = MemorySegment.NULL;
                freed = true;
            }
            if (failure != null) {
                throw failure;
            }
        }
    }

    void pageFreed(Page page) {
        // Called with the shared reentrant monitor held by Page.close().
        pages.remove(page);
    }

    private void requireOpen() {
        if (freed) {
            throw new IllegalStateException("browser is closed");
        }
    }

    private static RuntimeException combine(RuntimeException first, RuntimeException next) {
        if (first == null) {
            return next;
        }
        first.addSuppressed(next);
        return first;
    }
}
